eav/queryset.py

```diff
@@ -1,47 +1,273 @@
+            nargs.append(arg)
+
+        for key, value in kwargs.items():
+            # Modify kwargs (warning: recursion ahead).
+            nkey, nval = expand_eav_filter(self.model, key, value)
+
+            if nkey in nkwargs:
+                # Apply AND to both querysets.
+                nkwargs[nkey] = (nkwargs[nkey] & nval).distinct()
+            else:
+                nkwargs.update({nkey: nval})
+
+        return func(self, *nargs, **nkwargs)
+
+    return wrapper
+
+
+def expand_q_filters(q, root_cls):
+    '''
+    Takes a Q object and a model class.
+    Recursivley passes each filter / value in the Q object tree leaf nodes
+    through expand_eav_filter
+    '''
+    new_children = []
+
+    for qi in q.children:
+        if type(qi) is tuple:
+            # This child is a leaf node: in Q this is a 2-tuple of:
+            # (filter parameter, value).
+            key, value = expand_eav_filter(root_cls, *qi)
+            new_children.append(Q(**{key: value}))
+        else:
+            # This child is another Q node: recursify!
+            new_children.append(expand_q_filters(qi, root_cls))
+
+    q.children = new_children
+    return q
+
+
+def expand_eav_filter(model_cls, key, value):
+    '''
+    Accepts a model class and a key, value.
+    Recurisively replaces any eav filter with a subquery.
+
+    For example::
+
+        key = 'eav__height'
+        value = 5
+
+    Would return::
+
+        key = 'eav_values__in'
+        value = Values.objects.filter(value_int=5, attribute__slug='height')
+    '''
+    fields = key.split('__')
+    config_cls = getattr(model_cls, '_eav_config_cls', None)
+
+    if len(fields) > 1 and config_cls and fields[0] == config_cls.eav_attr:
+        slug = fields[1]
+        gr_name = config_cls.generic_relation_attr
+        datatype = Attribute.objects.get(slug=slug).datatype
+
+        lookup = '__%s' % fields[2] if len(fields) > 2 else ''
+        kwargs = {
+            'value_{}{}'.format(datatype, lookup): value,
+            'attribute__slug': slug
+        }
+        value = Value.objects.filter(**kwargs)
+
+        return '%s__in' % gr_name, value
+
+    try:
+        field = model_cls._meta.get_field(fields[0])
+    except models.FieldDoesNotExist:
+        return key, value
+
+    if not field.auto_created or field.concrete:
+        return key, value
+    else:
+        sub_key = '__'.join(fields[1:])
+        key, value = expand_eav_filter(field.model, sub_key, value)
+        return '{}__{}'.format(ields[0], key), value
 
 
 class EavQuerySet(QuerySet):
+    '''
+    Overrides relational operators for EAV models.
+    '''
+
     @eav_filter
     def filter(self, *args, **kwargs):
         '''
-        Pass *args* and *kwargs* through :func:`eav_filter`, then pass to
+        Pass *args* and *kwargs* through ``eav_filter``, then pass to
         the ``models.Manager`` filter method.
         '''
         return super().filter(*args, **kwargs).distinct()
```

Comment:
@IwoHerka is `distinct()` here valid and necessary?

Reply:
I'm not 100% sure. This is a left-over from the old codebase. I think it should be tested for corner-cases and then decided. However, I think it is beyond the scope of this pull-request (I'm trying not to dig into the legacy code too much) and I would like to investigate it in separation.

Reply:
I'll leave you a separate issue for that then.